### ISSUE TITLE
pytriqs.magic broken in [unstable]?

### ISSUE THREAD

**Issue by tayral** [TRIQS/triqs#400]
Hi,
Using the "unstable" branch, with g++ 7.0.1, I am encountering problems with pytriqs.magic in the jupyter notebook. I have a notebook with
```python
%reload_ext pytriqs.magic
```
and the cell:
```python
%%triqs             
#include <triqs/gfs.hpp>
using namespace triqs;
using namespace triqs::gfs;
        
gf<imfreq> demo(double beta) {
 int n_freq = 1000;
}
```
I get the warnings and error:
```bash
-----------------------------make error -----------------------------
Scanning dependencies of target ext
[ 50%] Building CXX object CMakeFiles/ext.dir/ext_wrap.cpp.o
In file included from /usr/include/python2.7/Python.h:85:0,
                 from /home/tayral/Work/TRIQS_1.6/install_gcc7/include/triqs/utility/first_include.hpp:62,
                 from /home/tayral/.cache/ipython/triqs/_triqs_magic_6ed112d84be3a88cdf0f74a3c88b1e29/ext_wrap.cpp:3:
/usr/include/python2.7/unicodeobject.h:534:24: warning: ISO C++1z does not allow ‘register’ storage class specifier [-Wregister]
     register PyObject *obj,     /* Object */
                        ^~~
/usr/include/python2.7/unicodeobject.h:553:24: warning: ISO C++1z does not allow ‘register’ storage class specifier [-Wregister]
     register PyObject *obj      /* Object */
                        ^~~
/usr/include/python2.7/unicodeobject.h:575:29: warning: ISO C++1z does not allow ‘register’ storage class specifier [-Wregister]
     register const wchar_t *w,  /* wchar_t buffer */
                             ^
/usr/include/python2.7/unicodeobject.h:593:23: warning: ISO C++1z does not allow ‘register’ storage class specifier [-Wregister]
     register wchar_t *w,        /* wchar_t buffer */
                       ^
In file included from /usr/include/python2.7/Python.h:94:0,
                 from /home/tayral/Work/TRIQS_1.6/install_gcc7/include/triqs/utility/first_include.hpp:62,
                 from /home/tayral/.cache/ipython/triqs/_triqs_magic_6ed112d84be3a88cdf0f74a3c88b1e29/ext_wrap.cpp:3:
/usr/include/python2.7/stringobject.h:173:24: warning: ISO C++1z does not allow ‘register’ storage class specifier [-Wregister]
     register PyObject *obj, /* string or Unicode object */
                        ^~~
/usr/include/python2.7/stringobject.h:174:21: warning: ISO C++1z does not allow ‘register’ storage class specifier [-Wregister]
     register char **s,  /* pointer to buffer variable */
                     ^
/usr/include/python2.7/stringobject.h:175:26: warning: ISO C++1z does not allow ‘register’ storage class specifier [-Wregister]
     register Py_ssize_t *len /* pointer to length variable or NULL
                          ^~~
In file included from /home/tayral/.cache/ipython/triqs/_triqs_magic_6ed112d84be3a88cdf0f74a3c88b1e29/ext_wrap.cpp:5:0:
/home/tayral/Work/TRIQS_1.6/install_gcc7/include/triqs/py_converters/gf.hpp:28:19: error: redefinition of ‘struct triqs::py_tools::py_converter<triqs::gfs::gf_mesh<triqs::gfs::imfreq> >’
 template<> struct py_converter<triqs::gfs::gf_mesh<triqs::gfs::imfreq>> {
                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /home/tayral/Work/TRIQS_1.6/install_gcc7/include/triqs/python_tools/converters/gf.hpp:8:0,
                 from /home/tayral/Work/TRIQS_1.6/install_gcc7/include/triqs/py_converters/gf.hpp:21,
                 from /home/tayral/.cache/ipython/triqs/_triqs_magic_6ed112d84be3a88cdf0f74a3c88b1e29/ext_wrap.cpp:5:
/home/tayral/Work/TRIQS_1.6/install_gcc7/include/triqs/py_converters/meshes.hpp:25:19: note: previous definition of ‘struct triqs::py_tools::py_converter<triqs::gfs::gf_mesh<triqs::gfs::imfreq> >’
 template<> struct py_converter<triqs::gfs::gf_mesh<triqs::gfs::imfreq>> {
(...)
```
Thomas

**Comment by tayral** [comment]
... after cleaning up, I get a different error:
```
-----------------------Wrapper generator error -----------------------
Traceback (most recent call last):
  File "ext_desc.py", line 9, in <module>
    module.use_module('gf', 'triqs')
  File "/home/tayral/Work/TRIQS_1.6/install_gcc7/share/triqs/cpp2py/wrap_generator.py", line 882, in use_module
    if not f : raise RuntimeError, "Cannot find the module %s.\n  ... module_path_list = %s"%(modulename, self.module_path_list)
RuntimeError: Cannot find the module gf.
  ... module_path_list = ['/home/tayral/Work/TRIQS_1.6/install_gcc7/include/triqs/py_converters/']

---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-3-82161f60cafb> in <module>()
----> 1 get_ipython().run_cell_magic(u'triqs', u'            ', u'#include <triqs/gfs.hpp>\nusing namespace triqs;\nusing namespace triqs::gfs;\n        \ngf<imfreq> demo(double beta) {\n\n int n_freq = 1000;\n\n \n}')

/usr/local/lib/python2.7/dist-packages/ipython-5.1.0-py2.7.egg/IPython/core/interactiveshell.pyc in run_cell_magic(self, magic_name, line, cell)
   2113             magic_arg_s = self.var_expand(line, stack_depth)
   2114             with self.builtin_trap:
-> 2115                 result = fn(magic_arg_s, cell)
   2116             return result
   2117 

<decorator-gen-122> in triqs(self, line, cell)

/usr/local/lib/python2.7/dist-packages/ipython-5.1.0-py2.7.egg/IPython/core/magic.pyc in <lambda>(f, *a, **k)
    186     # but it's overkill for just that one bit of state.
    187     def magic_deco(arg):
--> 188         call = lambda f, *a, **k: f(*a, **k)
    189 
    190         if callable(arg):

/home/tayral/Work/TRIQS_1.6/install_gcc7/lib/python2.7/dist-packages/pytriqs/magic.py in triqs(self, line, cell)
    149             execute ("""
    150             PYTHONPATH={generator_path} python ext_desc.py {generator_path}/mako/xxx_wrap.cpp ext_wrap.cpp {generator_path}/mako/py_converter.hpp conv.hpp {triqs_path}
--> 151             """.format(**globals()), "Wrapper generator")
    152 
    153             # Call cmake

/home/tayral/Work/TRIQS_1.6/install_gcc7/lib/python2.7/dist-packages/pytriqs/magic.py in execute(command, message)
    137                 except subprocess.CalledProcessError as E :
    138                    print_out ( message + " error ", E.output)
--> 139                    raise RuntimeError, "Error"
    140                 if args.verbosity>0:
    141                    print_out(message, out)

RuntimeError: Error

```

**Comment by tayral** [comment]
OK, this solves the issue with the simple script. However, with a function returning a lattice function (G(k,om)), the kernel crashes (with no error message):
```c++
%%triqs             
#include <triqs/gfs.hpp>
using namespace triqs;
using namespace triqs::gfs;
using triqs::clef::placeholder;
using triqs::clef::placeholder_prime;
gf<cartesian_product<brillouin_zone, imfreq>, matrix_valued> build_G0(){
    
 int nw      = 40;
 int nk      = 6;
 auto bz     = brillouin_zone{bravais_lattice{{{1., 0.}, {0.5, sqrt(3) / 2.}}}};
 double beta = 5, mu = 0.5;
 placeholder<0> k_;
 placeholder<4> inu_;
 auto eps_k_ = -2 * (cos(k_(0)) + cos(k_(1)));
 auto G_k_w  = gf<cartesian_product<brillouin_zone, imfreq>, matrix_valued>{{{bz, nk}, {beta, Fermion, nw}}, {1, 1}};
 G_k_w(k_, inu_) << 1.0 / (inu_ + mu - eps_k_);
 
 return G_k_w;
}
```
The above code compiles, but when calling build_G0, the notebook crashes with no further error:
```python
from pytriqs.gf import *
G = build_G0()
```
1) To investigate if this comes from a problem with "cartesian_product", I tried the same thing with functions of two frequencies, but it compiles and runs:
```c++
gf<cartesian_product<imfreq, imfreq>, matrix_valued> build_L0(){
 double beta=1;
 placeholder<0> iw_;
 placeholder<4> inu_;
 auto Lambda = gf<cartesian_product<imfreq, imfreq>, matrix_valued>{{{beta, Fermion, 5}, {beta, Boson, 5}}, {1,1}};
 Lambda(iw_, inu_) << 1.0 / (iw_) / (iw_+inu_);
 return Lambda;
}
```
But here I have issues plotting this object:
```python
from pytriqs.gf import *
L = build_L0()
from pytriqs.plot.mpl_interface import *
%pylab inline
oplot(L[[0,all]]) #plot at fixed first freq, vs. the second freq
```
returns the error:
```bash
TypeError                                 Traceback (most recent call last)
<ipython-input-17-7287fe50cdcf> in <module>()
      3 from pytriqs.plot.mpl_interface import *
      4 get_ipython().magic(u'pylab inline')
----> 5 oplot(L[[0,all]])

/home/tayral/Work/TRIQS_1.6/install_gcc7/lib/python2.7/dist-packages/pytriqs/gf/gf.pyc in __getitem__(self, key)
    309             dat = self._data[k]
    310             # list of the remaining lists
--> 311             mlist = [m for i,m in itertools.ifilter(lambda i,m : i is all, itertools.izip(key, mlist))]
    312             # if no mlist, then we have a complete evaluation
    313             # assert len(mlist) > 0, "Oops"  # More conservative version

TypeError: <lambda>() takes exactly 2 arguments (1 given)
```
2) To investigate if this is a problem with gf<brillouin_zone>, I tried using this type in a magic:
```c++
%%triqs             
#include <triqs/gfs.hpp>
using namespace triqs;
using namespace triqs::gfs;
using triqs::clef::placeholder;
using triqs::clef::placeholder_prime;
gf<brillouin_zone, matrix_valued> build_eps0(){
    
 int nk      = 6;
 auto bz     = brillouin_zone{bravais_lattice{{{1., 0.}, {0.5, sqrt(3) / 2.}}}};
 placeholder<0> k_;
 auto eps_k  = gf<brillouin_zone, matrix_valued>{{bz, nk}, {1, 1}};
 eps_k(k_) << -2 * (cos(k_(0)) + cos(k_(1)));
 
 return eps_k;
}
```
... but it compiles and runs (although I have the same problem with plotting as before, although this worked on the dev branch:
```python
from pytriqs.gf import *
eps_k = build_eps0()
oplot(eps_k)
```
returns
```bash
TypeError                                 Traceback (most recent call last)
<ipython-input-33-b7f8906b0b3c> in <module>()
      3 eps_k = build_eps0()
      4 
----> 5 oplot(eps_k)

/home/tayral/Work/TRIQS_1.6/install_gcc7/lib/python2.7/dist-packages/pytriqs/plot/mpl_interface.pyc in oplot(obj, *opt_list, **opt_dict)
     42     """
     43     plt.figure(num=opt_dict.pop('num', plt.gcf().number))
---> 44     __oplot_impl(plt, plt.xlabel, plt.ylabel, plt.legend, obj, plt.xticks, plt.title, *opt_list, **opt_dict)
     45 
     46 

/home/tayral/Work/TRIQS_1.6/install_gcc7/lib/python2.7/dist-packages/pytriqs/plot/mpl_interface.pyc in __oplot_impl(top, xlabel_fct, ylabel_fct, legend_fct, obj, xticks_fct, title_fct, *opt_list, **opt_dict)
     71     """
     72 
---> 73     for curve_dict in plot_protocol_apply(obj, opt_dict, plt.xlim):
     74 
     75         import warnings

/home/tayral/Work/TRIQS_1.6/install_gcc7/lib/python2.7/dist-packages/pytriqs/plot/protocol.pyc in plot_protocol_apply(ob, opt_dict, xlims)
     56 
     57     if hasattr(ob, '_plot_'):
---> 58         return ob._plot_(opt_dict)
     59     elif callable(ob):
     60         n_points = opt_dict.pop('n_points', 100)

/home/tayral/Work/TRIQS_1.6/install_gcc7/lib/python2.7/dist-packages/pytriqs/gf/gf.pyc in _plot_(self, opt_dict)
    543     def _plot_(self, opt_dict):
    544         """ Implement the plot protocol"""
--> 545         return plot.dispatcher(self)(self, opt_dict)
    546 
    547     def x_data_view(self, x_window=None, flatten_y=False):

TypeError: 'NoneType' object is not callable
```
)
In conclusion, there are two main problems
- issue with using the magic with gf<cart_prod<brillouin_zone, imfreq>
- issue with plotting gf<cart_prod<imfreq, imfreq>> and gf<brillouin_zone> (and probably with  gf<cart_prod<brillouin_zone, imfreq>)

**Comment by HugoStrand** [comment]
This is a bug in the double bracket accessor please see my pull request #399 for a fix.

**Comment by parcollet** [comment]
Merged.

**Comment by tayral** [comment]
Hi Hugo and Olivier,
No, unfortunately this does not fix any of the problems:
- build_G0() compiles but crashes when executed (I would like to reproduce something like [this](https://triqs.ipht.cnrs.fr/1.x/tutorials/gfs/lattice_gfs_tutorial_python.html))
- build_eps0() compiles, but the plotting does not work (same error message)
- build_L0() compiles, but the plotting does not work:
```python
oplot(L[[0,all]])
```
yields
```bash
TypeError                             Traceback (most recent call last)
<ipython-input-11-2ec8e6bd6e05> in <module>()
      4 get_ipython().magic(u'pylab inline')
      5 #for x in L.mesh.components[0]:
----> 6 oplot(L[[0,all]])
      7 #print L[0,0,0]

/home/tayral/Work/TRIQS_1.6/install_gcc7/lib/python2.7/dist-packages/pytriqs/gf/gf.pyc in __getitem__(self, key)
    305             # slice the data
    306             _ = slice(0, None) # all
--> 307             k = [m.index_to_linear(i) if i is not all else _ for i,m in itertools.izip(key, mlist)]
    308             k += self._target_rank * [_]
    309             dat = self._data[k]

TypeError: Cannot convert to long
```
It would be nice to recover the functionalities shown e.g. in [the documentation](https://triqs.ipht.cnrs.fr/1.x/tutorials/gfs/multivar_gfs_tutorial_python.html)

**Comment by parcollet** [comment]
Assigned to me. 
Studied the crash, and fix is easy  : all copy functions must be protected against failure.
Moreover, the tail of g(k, om ) is not yet wrapped, hence the conversion can not work
In progress

**Comment by tayral** [comment]
Any progress here? (I would like to use this for an upcoming tutorial... otherwise I can use a former version of TRIQS, but I'd rather benefit from the latest developments ;-))
Thomas

**Comment by tayral** [comment]
Hi,
This issue is not fixed (even after commit 3a48db8)
The code is:
```c++
gf<cartesian_product<brillouin_zone, imfreq>, matrix_valued> build_G0(){
    
 int nw      = 40;
 int nk      = 6;
 auto bz     = brillouin_zone{bravais_lattice{{{1., 0.}, {0.5, sqrt(3) / 2.}}}};
 double beta = 5, mu = 0.5;
 placeholder<0> k_;
 placeholder<4> inu_;
 auto eps_k_ = -2 * (cos(k_(0)) + cos(k_(1)));
 auto G_k_w  = gf<cartesian_product<brillouin_zone, imfreq>, matrix_valued>{{{bz, nk}, {beta, Fermion, nw}}, {1, 1}};
 G_k_w(k_, inu_) << 1.0 / (inu_ + mu - eps_k_);
 
 return G_k_w;
}
```
The execution of the cell
```python
from pytriqs.gf import *
G = build_G0()
```
yields the error
```bash
AssertionError                            Traceback (most recent call last)
<ipython-input-4-2eb7d0b2fc71> in <module>()
      1 
      2 from pytriqs.gf import *
----> 3 G = build_G0()

/home/tayral/Work/TRIQS_1.6/install_gcc7/lib/python2.7/dist-packages/pytriqs/gf/gf.py in __init__(self, **kw)
    186             self.__check_invariants()
    187 
--> 188         delegate(self, **kw)
    189 
    190     def __check_invariants_tail(self):

/home/tayral/Work/TRIQS_1.6/install_gcc7/lib/python2.7/dist-packages/pytriqs/gf/gf.py in delegate(self, mesh, data, target_shape, singularity, indices, name, is_real, _singularity_maker)
    164             if self._indices is not None:
    165                 d,i =  self._data.shape[self._rank:], tuple(len(x) for x in self._indices.data)
--> 166                 assert (d == i), "Indices are of incorrect size. Data size is %s while indices size is %s"%(d,i)
    167             # Now indices are set, and are always a GfIndices object, with the
    168             # correct size

AssertionError: Indices are of incorrect size. Data size is (11, 1, 1) while indices size is (1, 1)
```
